lib/PuppeteerSharp/FrameManager.cs

```diff
@@ -67,25 +67,21 @@ public async Task<Response> NavigateFrameAsync(Frame frame, string url, Navigati
             var watcher = new NavigatorWatcher(_client, this, frame, _networkManager, timeout, options);
 
             var navigateTask = NavigateAsync(_client, url, referrer, frame.Id);
-            await Task.WhenAny(
-                watcher.TimeoutOrTerminationTask,
-                navigateTask).ConfigureAwait(false);
+            await Task.Run(() => navigateTask, watcher.TimedCancellationToken).ConfigureAwait(false);
 
-            AggregateException exception = null;
+            Exception exception = null;
             if (navigateTask.IsFaulted)
             {
                 exception = navigateTask.Exception;
             }
             else
             {
-                await Task.WhenAny(
-                    watcher.TimeoutOrTerminationTask,
-                    _ensureNewDocumentNavigation ? watcher.NewDocumentNavigationTask : watcher.SameDocumentNavigationTask
-                ).ConfigureAwait(false);
 
-                if (watcher.TimeoutOrTerminationTask.IsCompleted && watcher.TimeoutOrTerminationTask.Result.IsFaulted)
+                await Task.Run(() => _ensureNewDocumentNavigation ? watcher.NewDocumentNavigationTask : watcher.SameDocumentNavigationTask, watcher.TimedCancellationToken).ConfigureAwait(false);
```

Comment:
Maybe we can move all this ` _ensureNewDocumentNavigation ? watcher.NewDocumentNavigationTask : watcher.SameDocumentNavigationTask` to a variable, so it's more clear.